gems/aws-sdk-core/lib/aws-sdk-core/cbor/decoder.rb

```diff
@@ -0,0 +1,304 @@
+          when 24 then :simple_value
+          when 25 then :half
+          when 26 then :float
+          when 27 then :double
+          when 31 then :break_stop_code
+          else
+            :reserved_undefined
+          end
+        end
+      end
+
+      def read_break_stop_code
+        read_info
+        :break_stop_code
+      end
+
+      def read_integer
+        major_type, add_info = read_info
+
+        val = read_count(add_info)
+        case major_type
+        when 0 then val
+        when 1 then -1 - val
+        else
+          raise CborError,
+                "Expected Integer (0,1) got major type: #{major_type}"
+        end
+      end
+
+      def read_binary_string
+        _major_type, add_info = read_info
+        take(read_count(add_info)).force_encoding(Encoding::BINARY)
+      end
+
+      def read_string
+        _major_type, add_info = read_info
+        take(read_count(add_info)).force_encoding(Encoding::UTF_8)
+      end
+
+      # returns only the length of the array, caller must read the correct number of values after this
+      def read_array
+        _major_type, add_info = read_info
+        read_count(add_info)
+      end
+
+      # returns nothing but consumes and checks the type/info.
+      # Caller must keep reading until encountering the stop sequence
+      def read_start_indefinite_array
+        read_info
+      end
+
+      # returns nothing but consumes and checks the type/info.
+      # Caller must keep reading until encountering the stop sequence
+      def read_start_indefinite_map
+        read_info
+      end
+
+      # returns nothing but consumes and checks the type/info.
+      def read_end_indefinite_collection
+        read_info
+      end
+
+      # returns only the length of the array, caller must read the correct number of key value pairs after this
+      def read_map
+        _major_type, add_info = read_info
+        read_count(add_info)
+      end
+
+      # returns only the tag, caller must interpret the tag and read another value as appropriate
+      def read_tag
+        _major_type, add_info = read_info
+        read_count(add_info)
+      end
+
+      def read_reserved_undefined
+        read_info
+      end
+
+      def read_boolean
+        _major_type, add_info = read_info
+        case add_info
+        when 20 then false
+        when 21 then true
+        else
+          raise CborError,
+                'Invalid Boolean simple type, expected add_info of 20 or 21, ' \
+                 "got: #{add_info}"
+        end
+      end
+
+      def read_nil
+        read_info
+        nil
+      end
+
+      def read_undefined
+        read_info
+        :undefined
+      end
+
+      # 16 bit IEEE 754 half-precision floats
+      # Support decoding only
+      # format:
+      # sign - 1 bit
+      # exponent - 5 bits
+      # precision - 10 bits
+      def read_half
+        read_info
+        b16 = take(2).unpack1('n')
```

Comment:
I know this is only used here, but its a lot of complexity to stick in an already hugely complex class.  I'd still potentially lean towards keeping it as a separate class